marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java

```diff
@@ -114,4 +113,68 @@ private String convertRowToJSONString(InternalRow record) {
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to parse ingestendpointparams, cause: " + e.getMessage(), e);
+        }
+
+        boolean doesNotHaveApiPath = !endpointParamsJsonNode.has("apiPath");
+        boolean hasWorkUnitOrEndpointState = endpointParamsJsonNode.has("workUnit") || endpointParamsJsonNode.has("endpointState");
+        if (doesNotHaveApiPath && hasWorkUnitOrEndpointState) {
+            throw new RuntimeException("Cannot override default endpointState and/or workUnit. Please provide custom endpoint.");
+        }
+
+        final String apiModulePath = (endpointParamsJsonNode.hasNonNull("apiPath") &&
+            endpointParamsJsonNode.get("apiPath").asText().length() > 0) ?
+            endpointParamsJsonNode.get("apiPath").asText() : "/data-hub/5/data-services/ingestion/bulkIngester.api";
+        logger.info("Will write to endpoint defined by: " + apiModulePath);
+        ((ObjectNode)endpointParamsJsonNode).put("apiPath", apiModulePath);
+
+        if (endpointParamsJsonNode.hasNonNull("endpointState") && endpointParamsJsonNode.get("endpointState").asText().length() > 0) {
+            ((ObjectNode)endpointParamsJsonNode).put("endpointState", endpointParamsJsonNode.get("endpointState").asText());
+        }
+        // TODO : remove the below else block after java-client-api 5.3 release
+        else {
+            ((ObjectNode)endpointParamsJsonNode).put("endpointState", "{}");
+        }
+
+        if (endpointParamsJsonNode.hasNonNull("workUnit") && endpointParamsJsonNode.get("workUnit").asText().length() > 0) {
+
+            JsonNode workUnitNode;
+            try {
+                workUnitNode = objectMapper.readValue(endpointParamsJsonNode.get("workUnit").asText(), JsonNode.class);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException("Unable to parse workUnit, cause: " + e.getMessage(), e);
+            }
+            ((ObjectNode) workUnitNode).put("uriprefix", uriPrefix);
+            ((ObjectNode)endpointParamsJsonNode).put("workUnit", workUnitNode.toString());
+        } else {
+            ((ObjectNode)endpointParamsJsonNode).put("workUnit", "{\"uriprefix\":\"" + uriPrefix + "\"}");
+        }
+        return endpointParamsJsonNode;
+    }
+
+    private InputEndpoint.BulkInputCaller buildBulkInputCaller(HubClient hubClient, String apiPath, String endpointState,
+                                                               String workUnit) {
+        InputEndpoint.BulkInputCaller bulkInputCaller;
+        try {
+            bulkInputCaller= InputEndpoint.on(
+                hubClient.getStagingClient(),
+                hubClient.getModulesClient().newJSONDocumentManager().read(apiPath, new StringHandle())
+            ).bulkCaller();
+        } catch (ResourceNotFoundException ex) {
+            throw new RuntimeException("Endpoint not found.");
```

Comment:
I am wondering if the ResourceNotFoundException needs to be caught here. If we do throw a new exception, we definitely need to include the original exception so that Ernie doesn't lose any context. 

I did a quick test, and this is what the exception shows:

```
com.marklogic.client.ResourceNotFoundException: Local message: Could not read non-existent document. Server Message: RESTAPI-NODOCUMENT: (err:FOER0000) Resource or document does not exist:  category: content message: /data-hub/5/data-services/ingestion/bulkIngesterrrr.api
```

That is likely helpful enough for Ernie - it shows the module he referenced, and the Java Client is giving a good indication of the error. So I think this try/catch can be removed. 

And because of that, we don't really need this method anymore, since it's just a few lines of code now. If we do need some special error handling here, I think it's worth the separate method. But that doesn't appear to be the case.